Add Err() and Close() methods to Iterators

andrew-d opened the pull request:
This PR does the following:
- As mentioned in #208, add an `Err()` method to all Iterators.
- Adds tests that the various iterators pass through the underlying Err value.
- Makes the `Close()` method on Iterators return an error.
- Remove a bunch of uses of `glog.Fatal*` calls, instead either propagating the errors, or calling `panic` (when it's a programming bug).
- Try to handle errors in other locations that use `Next()`.

cc @kortschak 

Updates #157

kortschak commented:
I had planned on adding Err() to Iterator, but this looks OK too - I can't really distinguish which is better. Can you provide pros for this approach? (On the face of it, this looks a bit neater in how it will pan out, but I'd like to hear your motivation here). Thanks for picking this up BTW.

Will look in more detail later.

andrew-d commented:
Well, the original reason that I added `Err()` to `Nexter` was mostly because I assumed the only reason you'd call it is when iterating using `Next()`.  However, if there are other things that might need it, I'd assume that I can just move it over to the `Iterator` interface.  This PR is still rather work-in-progress :smile:

kortschak commented:
Yeah. I think that's a plan. We can move it without concern if needed later. This does localise the feature to where it's needed AFAICS at the moment - when we dive into errors we may need expanding.

The Close method on graph.Iterator should return an error I think. Would you add that please.

andrew-d commented:
Okay, so, I've got most of the major iterators fixed, one per commit.  I only have two more to finish, and I'd appreciate some input:

**Materialize**  
Essentially, this runs the entire subiterator and caches the result.  Should this return an error if the subiterator returns one (i.e. return no results), or should it iterate over the results it has cached and then return the error?

**LinksTo**  
I'm not sure I understand what this iterator does, so I'm a bit leery of just randomly throwing error handling in.

And finally, a more general question - do you think it's worth having the various iterators check if the current error is non-nil in `Next()` and, if so, returning false?  The alternative being, of course, telling the user that "you must check for non-nil Err if Next returns false".

And finally, sure, I can add the error return to `Close()`.

kortschak commented:
My view is that an Err() != nil state means the Next call failed, and the user should find out why (this is provoked by not getting anything back). This would be consistent with Materialize returning no results on failure.

The justification for this is that partial results may not be safe, and we shouldn't be fostering unsafe behaviour.

Wait for @barakmich to give his view though.

andrew-d commented:
@kortschak Okay, added the `Close()` method.  One thing that is worth noting is that, for iterators that aggregate from underlying iterators (e.g. `Or`), if an error occurs, only the first one is returned.  I did my best to document this, but thoughts are appreciated.

kortschak commented:
I think it's reasonable to only return the first. We can extend that at a later date if it's deemed necessary - I guess it really depends on the needs of the backend (the authors of those may want to add handling for multiple error returns via a []error type that implements error).

barakmich commented:
So I'd say that it's probably wise to put Err() on the iterator interface. A good example of this is HasA.Check() which can Next() its way through a subiterator and possibly have an error therein. The good news is it's a slight interface change away. :)

I'm +1 with @kortschak that partial results may not be safe, and consistent with Materialize returning no results on failure. Not to conflate it with no results on success. ;) 

The rule of thumb being, as an iterator, if a subiterator has an error, we should catch it and error ourselves. Whatever that first error happens to be (extending to []error if necessary, as discussed). Materialize is actually simple, because it's exactly this -- the only time it produces errors of its own should be if memory allocation fails or something equivalently implementation-detail-oriented, because it's the same set of things as the subiterator, just pre-fetched.

LinksTo is, to explain a bit of what it does, the set of all links that are "links to" the subiterator of nodes. HasA is the set of nodes that a subiterator of links has a specific field of. (`foo bar baz .` "has a" subject being `foo`)

Also, thank you so much for digging in with both hands. This is awesome and cleans up a lot of the code :+1:

andrew-d commented:
@barakmich @kortschak Okay, I've added all the various Err / Close methods.  It does appear that I've broken the tests somewhat - `go test -v ./query/gremlin` is a simple example of what no longer works.  I've tried to avoid any functional changes, but clearly I missed something :smile: 

I'll go through the code again sometime today, but if you have any suggestions, it would be appreciated.

andrew-d commented:
Aha!  Okay, that last commit fixed it.  `git bisect` is great :smiley: 

So: all tests pass, all `Iterator`s have an `Err()` method, and all `Close()` calls return an error.  @barakmich, the last question I have for you relates to error handling in other parts of the code.  Currently, I more or less assumed that errors really only occur during iteration (i.e. during `Next()`).  Is there any other parts of the code you want me to fiddle with?

kortschak commented:
Have a look for any glog.Fatal calls. Function that contain them, should return an error if they don't already.

barakmich commented:
And take a look for errors related to NextResult() and Check() -- those are iteration-related.

andrew-d commented:
@kortschak Will do.

@barakmich So, maybe I'm missing something, but I don't see any functions called either `NextResult` or `Check` in the code?

andrew-d commented:
@kortschak, @barakmich: Okay, I removed a bunch of `glog.Fatal*` calls.  I converted a couple to `panic`s - stuff that looked like a programming bug, not necessarily an error.

Also: I added more error handling in the HasA iterator.  I think that, as of now, all uses of `Next()` in the code are either:
- In `Next()` methods themselves
- In tests
- Or properly handle errors.

Feel free to double-check - here's what I've been using: `ag '\.Next\(' -G '^(?!(.*_test\.go)$)'`

kortschak commented:
NextResult and Check got renamed a while ago to NextPath and (from dim memory) Contains.

barakmich commented:
Correct. Old dog, new tricks. ;)

On 4/15/2015 5:25 PM, Dan Kortschak wrote:
> NextResult and Check got renamed a while ago to NextPath and (from dim 
> memory) Contains.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/google/cayley/pull/229#issuecomment-93573776>.
>

andrew-d commented:
@barakmich Done!  That being said, I noticed [something strange](https://github.com/google/cayley/pull/229/files#diff-aff85d12107d6522135107af61084319R199) - any idea if that result is supposed to be used?

kortschak commented on graph/iterator.go:
```diff
@@ -100,6 +100,9 @@ type Iterator interface {
 	// Contains returns whether the value is within the set held by the iterator.
 	Contains(Value) bool
 
+	// Err returns the error (if any) encountered during iteration.
```
Err returns any error that was encountered by the Iterator.

kortschak commented on graph/iterator/all_iterator.go:
```diff
@@ -103,6 +105,11 @@ func (it *Int64) Next() bool {
 }
 
+func (it *Int64) Err() error {
+	// This iterator should never error.
```
I don't think this comment is necessary.

kortschak commented on graph/iterator/and_iterator.go:
```diff
@@ -255,13 +276,25 @@ func (it *And) cleanUp() {}
 // follow this contract, the And follows the contract.
-func (it *And) Close() {
+//
+// Note: as this potentially involves closing multiple subiterators, only
```
Don't add this as a note, just say "It closes all sub iterators it can, but returns the first error it encounters." (cribbed from os.RemoveAll).

Also below where this is used.

kortschak commented on graph/iterator/fixed_iterator.go:
```diff
@@ -143,6 +145,11 @@ func (it *Fixed) Next() bool {
 }
 
+func (it *Fixed) Err() error {
+	// This iterator should never error.
```
Drop comment.

kortschak commented on graph/iterator/hasa_iterator.go:
```diff
@@ -239,11 +262,20 @@ func (it *HasA) Stats() graph.IteratorStats {
 // Close the subiterator, the result iterator (if any) and the HasA.
-func (it *HasA) Close() {
+//
+// Note: as this involves closing multiple iterators, only the first error
```
See above.

barakmich commented:
@andrew-d That strange bit should really have a comment: 

```
// One of the iterators has determined that this value doesn't match. 
// However, the iterators that came before in the list may have returned "ok" to Contains(). 
// We need to set all the tags back to what the previous result was -- effectively 
// seeking back exactly one -- so we check all the prior iterators with the (already verified) result and 
// throw away the result, which will be 'true'
```

And check them for errors in that list. 

Well caught!

kortschak commented on graph/iterator/hasa_iterator_test.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+func TestHasAIteratorErr(t *testing.T) {
+	retErr := errors.New("unique")
```
s/retErr/wantErr/g

kortschak commented on graph/iterator/linksto_iterator.go:
```diff
@@ -176,19 +189,38 @@ func (it *LinksTo) Next() bool {
-	it.nextIt.Close()
-	it.primaryIt.Close()
+//
+// Note: as this involves closing multiple subiterators, only the first error
```
See above.

kortschak commented on graph/iterator/materialize_iterator_test.go:
```diff
@@ -0,0 +1,72 @@
+	"errors"
+	"testing"
+
+	//"github.com/google/cayley/graph"
```
Delete.

kortschak commented on graph/iterator/materialize_iterator_test.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+func TestMaterializeIteratorError(t *testing.T) {
+	retErr := errors.New("unique")
```
s/retErr/wantErr/g

kortschak commented on graph/iterator/materialize_iterator_test.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func TestMaterializeIteratorErrorAbort(t *testing.T) {
+	retErr := errors.New("unique")
```
As above.

kortschak commented on graph/iterator/materialize_iterator_test.go:
```diff
@@ -0,0 +1,72 @@
+
+	mIt := NewMaterialize(or)
+
+	// Should get all the underlying values...
```
We should...

kortschak commented on graph/iterator/materialize_iterator_test.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+
+	// ... and then the error value
```
Finish with a period.

kortschak commented on graph/iterator/not_iterator.go:
```diff
@@ -115,9 +130,19 @@ func (it *Not) NextPath() bool {
-func (it *Not) Close() {
-	it.primaryIt.Close()
-	it.allIt.Close()
+// Close closes the primary and all iterators.  If an error occurs, only the
```
Use same phrasing as above.

kortschak commented on graph/iterator/or_iterator.go:
```diff
@@ -231,12 +252,21 @@ func (it *Or) cleanUp() {}
-func (it *Or) Close() {
+// follow this contract, the Or follows the contract.
+//
+// Note: as this potentially involves closing multiple subiterators, only
```
See above.

kortschak commented on graph/iterator/or_iterator_test.go:
```diff
@@ -148,3 +149,46 @@ func TestShortCircuitingOrBasics(t *testing.T) {
 }
+
+func TestOrIteratorErr(t *testing.T) {
+	retErr := errors.New("unique")
```
s/retErr/wantErr/g

kortschak commented on graph/iterator/or_iterator_test.go:
```diff
@@ -148,3 +149,46 @@ func TestShortCircuitingOrBasics(t *testing.T) {
+}
+
+func TestShortCircuitOrIteratorErr(t *testing.T) {
+	retErr := errors.New("unique")
```
As above.

kortschak commented on graph/iterator/value_comparison_iterator.go:
```diff
@@ -168,7 +175,11 @@ func (it *Comparison) Contains(val graph.Value) bool {
 		return false
 	}
-	return it.subIt.Contains(val)
+	ret := it.subIt.Contains(val)
```
s/ret/ok/g

kortschak commented on graph/iterator/value_comparison_iterator_test.go:
```diff
@@ -118,3 +119,17 @@ func TestVCIContains(t *testing.T) {
 }
+
+func TestComparisonIteratorErr(t *testing.T) {
+	retErr := errors.New("unique")
```
s/retErr/wantErr/g

kortschak commented on graph/primarykey.go:
```diff
@@ -84,8 +84,9 @@ func (p *PrimaryKey) Int() int64 {
 	case unique:
-		glog.Fatal("UUID cannot be cast to an int64")
-		return -1
+		msg := "UUID cannot be cast to an int64"
```
s/cast/converted/

kortschak commented on writer/single.go:
```diff
@@ -33,17 +33,24 @@ type Single struct {
 
 func NewSingleReplication(qs graph.QuadStore, opts graph.Options) (graph.QuadWriter, error) {
 	var ignoreMissing, ignoreDuplicate bool
+	var err error
```
Use a var block.

kortschak commented:
I've done a first pass. In additions to the specific comments, three general comments:

1. Use `err` or `ok` (or other contextually meaningful name) in place of `ret`. To avoid shadowing where it happens because of this, rename temporary error values to something like `_err`.
2. Where you have `if expr; cond {` replace with `expr; if cond {` unless you really need the scoping or the readability if significantly improved (a good example of this is when you are doing a single map existance test), otherwise the conditional is harder to read.
3. Move err fields to bottom of the struct (this seems to be broadly conventional).

kortschak commented:
Oh, also, when this gets merged it should have "Updates #157." in the merge commit message.

andrew-d commented:
@barakmich, @kortschak: Okay, I think I addressed all your comments, and the message you mentioned is in the PR description (which I think means it gets into the merge commit).

kortschak commented on graph/bolt/iterator.go:
```diff
@@ -46,6 +46,7 @@ type Iterator struct {
 	dir     quad.Direction
 	qs      *QuadStore
 	result  *Token
+	err     error
```
Move this one down.

kortschak commented on graph/iterator/linksto_iterator.go:
```diff
@@ -176,19 +187,34 @@ func (it *LinksTo) Next() bool {
-func (it *LinksTo) Close() {
-	it.nextIt.Close()
-	it.primaryIt.Close()
+// Close our subiterators.  It closes all subiterators it can, but
```
Phrasing here.

andrew-d replied:
Do you mind clarifying?  I'm not seeing anything wrong here, but I might be missing something.

kortschak replied:
Sorry, clarified in edit.

kortschak commented:
That style is better thanks.

LGTM with minor nits.

andrew-d commented:
Okay, I've addressed all @kortschak's comments :smile:

Let me know if anything else needs to be done!

barakmich commented:
Just cause I've gotta ask: You've signed the CLA?

andrew-d commented:
If you mean [this](https://cla.developers.google.com/), I have now :smile_cat:

barakmich commented:
Excellent! Merging this, as it LGTM too with another look.

A good followup PR is adding yourself to the A+C, @andrew-d . :+1:
